web/src/client/storage.js

```diff
@@ -58,25 +60,140 @@ const removeUndefinedCockpitProperties = (cockpitObject) => {
+ */
+class DevicesManager {
+  /**
+   * @param {DBusClient} client
+   * @param {string} rootPath - Root path of the devices tree
+   */
+  constructor(client, rootPath) {
+    this.client = client;
+    this.rootPath = rootPath;
+  }
+
+  /**
+   * Gets all the exported devices
+   *
+   * @returns {Promise<StorageDevice[]>}
+   *
+   * @typedef {object} StorageDevice
+   * @property {string} sid - Internal id that is used as D-Bus object basename
+   * @property {string} type - Type of device ("disk", "raid", "multipath", "dasd", "md")
+   * @property {string} [vendor]
+   * @property {string} [model]
+   * @property {string[]} [driver]
+   * @property {string} [bus]
+   * @property {string} [transport]
+   * @property {boolean} [sdCard]
+   * @property {boolean} [dellBOOS]
+   * @property {string[]} [devices] - RAID devices (only for "raid" type)
+   * @property {string} [level] - MD RAID level (only for "md" type)
+   * @property {string} [uuid]
+   * @property {string[]} [members] - Member devices for a MD RAID (only for "md" type)
+   * @property {boolean} [active]
+   * @property {string} [name] - Block device name
+   * @property {number} [size]
+   * @property {string[]} [systems] - Name of the installed systems
+   * @property {string[]} [udevIds]
+   * @property {string[]} [udevPaths]
+   * @property {PartitionTableData} [partitionTable]
+   *
+   * @typedef {object} PartitionTableData
+   * @property {string} type
+   */
+  async getDevices() {
+    const buildDevice = (path, dbusDevice) => {
+      const addDriveProperties = (device, dbusProperties) => {
+        device.type = dbusProperties.Type.v;
+        device.vendor = dbusProperties.Vendor.v;
+        device.model = dbusProperties.Model.v;
+        device.driver = dbusProperties.Driver.v;
+        device.bus = dbusProperties.Bus.v;
+        device.transport = dbusProperties.Transport.v;
+        device.sdCard = dbusProperties.Info.v.SDCard.v;
+        device.dellBOSS = dbusProperties.Info.v.DellBOSS.v;
+      };
+
+      const addRAIDProperties = (device, raidProperties) => {
+        device.devices = raidProperties.Devices.v;
+      };
+
+      const addMDProperties = (device, mdProperties) => {
+        device.type = "md";
+        device.level = mdProperties.Level.v;
+        device.uuid = mdProperties.UUID.v;
+        device.members = mdProperties.Members.v;
+      };
+
+      const addBlockProperties = (device, blockProperties) => {
+        device.active = blockProperties.Active.v;
+        device.name = blockProperties.Name.v;
+        device.size = blockProperties.Size.v;
+        device.systems = blockProperties.Systems.v;
+        device.udevIds = blockProperties.UdevIds.v;
+        device.udevPaths = blockProperties.UdevPaths.v;
+      };
+
+      const addPtableProperties = (device, ptableProperties) => {
+        device.partitionTable = { type: ptableProperties.Type.v };
+      };
+
+      const device = {
+        sid: path.split("/").pop(),
```

Comment:
NP: [`pop()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/pop) is a mutating method, which make it less performant than simply using `[0]`

Reply:
Yes, but we need the last element. I guess that pop operation is cheap enough anyway.